### DISCUSSION TITLE
Reporting duplicated vulnerabilities on containers with SPDX files

### DISCUSSION THREAD

**Discussion by juan131** [aquasecurity/trivy#5446] [Bugs]
### Description

Trivy is reporting some duplicated detected vulnerabilities on images such as Bitnami ones that include a SPDX file. Please refer to the link below to understand how to detect these files:

- https://github.com/bitnami/vulndb#how-to-consume-this-cve-feed

These SPDX files include info about the packages included in the container image (e.g. JARs) and Trivy uses this info to detect vulnerabilities that might affect them. Check this example from Cassandra SPDX file:

```json
(...)
        {
            "name": "com.google.guava:guava",
            "SPDXID": "SPDXRef-Package-45b0907a54ea152b",
            "versionInfo": "27.0-jre",
            "supplier": "NOASSERTION",
            "downloadLocation": "NONE",
            "licenseConcluded": "NONE",
            "licenseDeclared": "NONE",
            "copyrightText": "NOASSERTION",
            "externalRefs": [
                {
                    "referenceCategory": "PACKAGE_MANAGER",
                    "referenceType": "purl",
                    "referenceLocator": "pkg:maven/com.google.guava/guava@27.0-jre"
                }
            ],
            "filesAnalyzed": false
        }
(...)
```

Trivy also has capabilities to autodetect JARs (and other packages) in the container filesystem and it seems it's unable to detect duplicates when the same package is both detected as a JAR and a SPDX package.

### Desired Behavior

Using `bitnami/cassandra` as example, it should report `CVE-2023-2976` just once:

```console
$ trivy image -q --scanners vuln --vuln-type library bitnami/cassandra -f json | jq -r '[.Results[].Vulnerabilities[] | select(.PkgName == "com.google.guava:guava" and .VulnerabilityID == "CVE-2023-2976")] | length'
1
```

### Actual Behavior

Using `bitnami/cassandra` as example, it reports `CVE-2023-2976` twice:

```console
$ trivy image -q --scanners vuln --vuln-type library bitnami/cassandra -f json | jq -r '[.Results[].Vulnerabilities[] | select(.PkgName == "com.google.guava:guava" and .VulnerabilityID == "CVE-2023-2976")] | length'
2
```

### Reproduction Steps

```bash
1. Run `trivy image` with any Bitnami image containing JARs vulnerabilities.
2. Look for duplicates in the the detected vulnerabilities.
```

### Target

Container Image

### Scanner

Vulnerability

### Output Format

JSON

### Mode

Standalone

### Debug Output

```bash
I don't think it adds any value.
```

### Operating System

macOS Ventura

### Version

```bash
$ trivy --version
Version: 0.46.0
Vulnerability DB:
  Version: 2
  UpdatedAt: 2023-10-25 12:18:46.676386342 +0000 UTC
  NextUpdate: 2023-10-25 18:18:46.676385542 +0000 UTC
  DownloadedAt: 2023-10-25 12:25:09.855557 +0000 UTC
Java DB:
  Version: 1
  UpdatedAt: 2023-10-25 00:53:36.560239655 +0000 UTC
  NextUpdate: 2023-10-28 00:53:36.560239155 +0000 UTC
  DownloadedAt: 2023-10-25 11:43:14.740794 +0000 UTC
```

### Checklist

- [X] Run `trivy image --reset`
- [X] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)

**Comment by juan131** [comment]
I applied this fix locally and it addressed the issue but I assume there are better ways to address it:
```diff
diff --git a/pkg/scanner/langpkg/scan.go b/pkg/scanner/langpkg/scan.go
index 9480c8a1..87fa9cdf 100644
--- a/pkg/scanner/langpkg/scan.go
+++ b/pkg/scanner/langpkg/scan.go
@@ -78,9 +78,24 @@ func (s *scanner) Scan(target types.ScanTarget, _ types.ScanOptions) (types.Resu
                        continue
                }

+               var uniqueVulns []types.DetectedVulnerability
+               for i := 0; i < len(vulns); i++ {
+                       duplicate := false
+                       for j := 0; j < len(uniqueVulns); j++ {
+                               if vulns[i].PkgName == uniqueVulns[j].PkgName &&
+                                       vulns[i].VulnerabilityID == uniqueVulns[j].VulnerabilityID {
+                                       duplicate = true
+                                       break
+                               }
+                       }
+                       if !duplicate {
+                               uniqueVulns = append(uniqueVulns, vulns[i])
+                       }
+               }
+
                results = append(results, types.Result{
                        Target:          targetName(app.Type, app.FilePath),
-                       Vulnerabilities: vulns,
+                       Vulnerabilities: uniqueVulns,
                        Class:           types.ClassLangPkg,
                        Type:            app.Type,
                })
@@ -88,6 +103,7 @@ func (s *scanner) Scan(target types.ScanTarget, _ types.ScanOptions) (types.Resu
        sort.Slice(results, func(i, j int) bool {
                return results[i].Target < results[j].Target
        })
+
        return results, nil
 }
```

**Reply by knqyf263** [reply]
Thanks for your detailed report! We try to deduplicate vulnerabilities here. @DmitriyLewen Can you please look into it?
https://github.com/aquasecurity/trivy/blob/57fa701a87baea1d8ffe4d68a311ad410024ce8a/pkg/result/filter.go#L110

**Reply by DmitriyLewen** [reply]
I am not sure that we need to filter vulnerabilities. 
Report includes one jar as 2 same packages with different `FilePath`:
```bash
➜ trivy image -q --scanners vuln --list-all-pkgs --vuln-type library bitnami/cassandra -f json | jq '.Results[].Packages[] | select(.Name == "com.google.guava:guava")'
{
  "Name": "com.google.guava:guava",
  "Version": "27.0-jre",
  "Layer": {
    "Digest": "sha256:5a5664f31ecd2833734f5c01cb8f872ebe0ba697b2c15c1252f7e8a92795c17c",
    "DiffID": "sha256:17598a690fe05a25c8cf5a9deec6267ecdeccae12e442377175445329547ef3d"
  },
  "FilePath": "opt/bitnami/cassandra/lib/guava-27.0-jre.jar"
}
{
  "Name": "com.google.guava:guava",
  "Version": "27.0-jre",
  "Ref": "pkg:maven/com.google.guava/guava@27.0-jre",
  "Layer": {
    "Digest": "sha256:5a5664f31ecd2833734f5c01cb8f872ebe0ba697b2c15c1252f7e8a92795c17c",
    "DiffID": "sha256:17598a690fe05a25c8cf5a9deec6267ecdeccae12e442377175445329547ef3d"
  },
  "FilePath": "opt/bitnami/cassandra"
}
```
I think we need to filter packages (keep only 1 package).

**Reply by knqyf263** [reply]
Yeah, that's why we respect the package path here.
https://github.com/aquasecurity/trivy/blob/57fa701a87baea1d8ffe4d68a311ad410024ce8a/pkg/result/filter.go#L133

@DmitriyLewen So, you think the result @juan131 shared is legitimate, right?

**Reply by DmitriyLewen** [reply]
> So, you think the result @juan131 shared is legitimate, right?

Right. Trivy found 2 packages and showed vulnerability for each package.

___

i investigated a bit:
we have logic to update filepath for packages from bitnami SBOM:
https://github.com/aquasecurity/trivy/blob/26b4959541b8fb10adb5f454ad930c6e2b68a0a8/pkg/fanal/analyzer/sbom/sbom.go#L89

But spdx file in `bitnami/cassandra` doesn't have information about filepaths:
```bash
9be2de2fe8c8:/opt/bitnami/cassandra$ cat /opt/bitnami/cassandra/.spdx-cassandra.spdx | grep "jar"
            "name": "jar",
```

If will add filepath in SPDX file - Trivy will not add duplicates ( i manually added filepath in Trivy code to show you):
```
➜  trivy git:(fix/cyclonedx-empty-purl) ✗ ./trivy image -q --scanners vuln --list-all-pkgs --vuln-type library bitnami/cassandra -f json | jq '.Results[].Packages[] | select(.Name == "com.google.guava:guava")'
{
  "Name": "com.google.guava:guava",
  "Version": "27.0-jre",
  "Ref": "pkg:maven/com.google.guava/guava@27.0-jre",
  "Layer": {
    "DiffID": "sha256:17598a690fe05a25c8cf5a9deec6267ecdeccae12e442377175445329547ef3d"
  },
  "FilePath": "opt/bitnami/cassandra/lib/guava-27.0-jre.jar"
}

```

**Reply by juan131** [reply]
Thanks so much for the detailed analysis @DmitriyLewen, the missing filepaths is a topic we already discussed in a couple of discussions in the past and there were some issue validating SPDX files with [SPDX validator](https://tools.spdx.org/app/validate), see:

- https://github.com/aquasecurity/trivy/pull/5062#discussion_r1311699709
- https://github.com/aquasecurity/trivy/discussions/4507

cc @gongomgra

With that said, shouldn't be enough with a match on VulnerabilityID, PkgName & InstalledVersion? I'm talking about the condition below:

- https://github.com/aquasecurity/trivy/blob/57fa701a87baea1d8ffe4d68a311ad410024ce8a/pkg/result/filter.go#L133

**Reply by DmitriyLewen** [reply]
We think that we should include in report vulnerabilities for each package.
If we remove `vuln.PkgPath` for `key` - user will only see 1 vulnerable package (as in https://github.com/aquasecurity/trivy/issues/3297).

But i still think we need to remove package duplicates for your case.

But i have 1 more thought:
You can create [module](https://aquasecurity.github.io/trivy/v0.46/docs/advanced/modules/) to remove duplicate vulnerabilities and we (if this is acceptable for @knqyf263 ) will add this module to [docs ecosystem](https://aquasecurity.github.io/trivy/v0.46/ecosystem/)
@knqyf263 wdyt?

**Reply by knqyf263** [reply]
As @DmitriyLewen shared, the same packages can be installed in different paths. We cannot drop the package path condition as Trivy should report all of them. 
Ideally, we should include the package path in SPDX somehow.

**Reply by juan131** [reply]
I think that including the correct "PkgPath" info would be better than creating a module. I'm syncing with @gomgomra trying to identify what we need to report the "file path" info in Bitnami SPDX files

**Reply by knqyf263** [reply]
Yes, that's true. We're here to help you improve file paths in SPDX. Please feel free to ask anything.

**Reply by DmitriyLewen** [reply]
@juan131 I'm working on adding `FilesAnalyzed` and `PackageVerificationCode` - https://github.com/aquasecurity/trivy/pull/5533
Perhaps these changes will help you.

**Reply by juan131** [reply]
That'll definitely help us @DmitriyLewen ! Thanks

**Comment by juan131** [comment]
I'm glad to announce that we have adapted our SPDX generation logic to properly filled the missing `files[]` & `relationships` information thanks to the enhancements @DmitriyLewen included on Trivy `0.48.0`.

We just released a new version of Cassandra container (`bitnami/cassandra:4.0.11-debian-11-r85`) with thes new improved SPDX files. As you can see, duplicates on reported vulnerabilities are gone:

- Before:

```console
$ trivy i bitnami/cassandra:4.0.11-debian-11-r84 --quiet --vuln-type library -f json | jq -r '.Results[].Vulnerabilities[] | [.PkgName, .PkgPath, .PkgRef, .VulnerabilityID] | @tsv' | grep guava
com.google.guava:guava	opt/bitnami/cassandra	pkg:maven/com.google.guava/guava@27.0-jre	CVE-2023-2976
com.google.guava:guava	opt/bitnami/cassandra/lib/guava-27.0-jre.jar		CVE-2023-2976
com.google.guava:guava	opt/bitnami/cassandra	pkg:maven/com.google.guava/guava@27.0-jre	CVE-2020-8908
com.google.guava:guava	opt/bitnami/cassandra/lib/guava-27.0-jre.jar		CVE-2020-8908
```

- After:

```console
$ trivy i bitnami/cassandra:4.0.11-debian-11-r85 --quiet --vuln-type library -f json | jq -r '.Results[].Vulnerabilities[] | [.PkgName, .PkgPath, .PkgRef, .VulnerabilityID] | @tsv' | grep guava
com.google.guava:guava	opt/bitnami/cassandra/lib/guava-27.0-jre.jar	pkg:maven/com.google.guava/guava@27.0-jre	CVE-2023-2976
com.google.guava:guava	opt/bitnami/cassandra/lib/guava-27.0-jre.jar	pkg:maven/com.google.guava/guava@27.0-jre	CVE-2020-8908
```
